lib/cisco_node_utils/cmd_ref/README_YAML.md

```diff
@@ -0,0 +1,458 @@
+# Command Reference YAML
+
+The [YAML](http://yaml.org) files in this directory are used with the
+`Cisco::CommandReference` module as a way to abstract away differences
+between client APIs as well as differences between platforms sharing
+the same client API.
+
+This document describes the structure and semantics of these files.
+
+* [Introduction](#introduction)
+* [Basic attribute definition](#basic-attribute-definition)
+  * [Wildcard substitution](#wildcard-substitution)
+    * [Printf-style wildcards](#printf-style-wildcards)
+    * [Key-value wildcards](#key-value-wildcards)
+* [Advanced attribute definition](#advanced-attribute-definition)
+  * [`_template`](#_template)
+  * [API variants](#api-variants)
+  * [Platform variants](#platform-variants)
+  * [Combinations of these](#combinations-of-these)
+* [Attribute properties](#attribute-properties)
+  * [`config_get`](#config_get)
+  * [`config_get_token`](#config_get_token)
+  * [`config_get_token_append`](#config_get_token_append)
+  * [`config_set`](#config_set)
+  * [`config_set_append`](#config_set_append)
+  * [`default_value`](#default_value)
+  * [`test_config_get` and `test_config_get_regex`](#test_config_get-and-test_config_get_regex)
+  * [`test_config_result`](#test_config_result)
+* [Style Guide](#style-guide)
+
+## Introduction
+
+Each file describes a single 'feature' (a closely related set of
+configurable attributes). The YAML within the file defines the set of
+'attributes' belonging to this feature. When a `CommandReference` object
+is instantiated, the user can look up any given attribute using the
+`lookup('feature_name', 'attribute_name')` API. Usually, instead of calling
+this API directly, node utility classes will call the various `Node` APIs,
+for example:
+
+```ruby
+config_set('feature_name', 'attribute_name', *args)
+value = config_get('feature_name', 'attribute_name')
+default = config_get_default('feature_name', 'attribute_name')
+```
+
+## Basic attribute definition
+
+The simplest definition of an attribute directly sets one or more properties
+of this attribute. These properties' values can generally be set to any
+basic Ruby type such as string, boolean, integer, array, or regexp.
+An example:
+
+```yaml
+# vtp.yaml
+domain:
+  config_get: "show vtp status"
+  config_get_token: "domain_name"
+  config_set: "vtp domain %s"
+
+filename:
+  config_get: "show running vtp"
+  config_get_token: '/vtp file (\S+)/'
+  config_set: "%s vtp password %s"
```

Comment:
Given that we are recommending wildcard substitution, I would prefer we use it in this and other examples.  Except the example showing Printf-style wildcards of course :)